File: z/包含min函数的栈.cc
class Solution {
 public:
  void push(int value) {
    s.push(value);
    if (helper.empty()) {
      helper.push(value);
    } else {
      if (helper.top() > value) {
        helper.push(value);
      }
    }
  }
  void pop() {
    if (s.top() > helper.top()) {
      s.pop();
    } else if (s.top() == helper.top()) {
      s.pop();
      helper.pop();
    }
  }
  int top() { return s.top(); }
  int min() { return helper.top(); }
  stack<int> s;
  stack<int> helper;
};

File: h/包含min函数的栈.java
import java.util.Stack;
import java.util.ArrayList;
public class Solution {
   private Stack<Integer> stack = new Stack<Integer>();
    private Stack<Integer> min = new Stack<Integer>();
     
    public void push(int node) {
        stack.push(node);
        if (min.isEmpty() || node < min.peek()) {
            min.push(node);
        }
    }
     
    public void pop() {
        if (stack.peek() == min.peek()) {
            min.pop();
        }
        stack.pop();
    }
     
    public int top() {
        return stack.peek();
    }
     
    public int min() {
        return min.peek();
    }
}
